### PR TITLE
feat: change calculation of protein_pool exchange UB

### PR DESCRIPTION
**Pull request by edkerk**
I've encountered cases in two different projects (two different yeasts) where the total protein content was drastically reduced in some experimental conditions. This low protein content was problematic when making conditoin-specific ec-models with proteomics data integrated. I've made the following changes to deal with this.

Two functions were changed to make them more flexible, with full backwards compatibility, just added functionality:
- `measureAbundance`: allow the use of provided arrays of pIDs and abundances, instead of only using the `prot_abundances` file. This is useful if one would like to calculate condition-specific `f`-factor. I have seen ec-models where the protein content was reduced by 50% in a particular growth condition, but not equally for enzymes and non-enzymes, resulting in a very different `f`-factor than from standard conditions. The changed `measureAbundance` is fully backwards compatible. Commit: fd1941214bb66cecbaf7fb53ea038c7802a1f145.

- `constrainEnzymes`: allow the use of a parameters structure as input, instead of only calling `getModelParameters` in the function. This is helpful in a script that loops through some code to generate multiple ec-models for different conditions (carbon sources). This would otherwise require multiple `getModelParameters` files that would have to be moved/renamed as part of the loop. Commit: ca2dd52ea016be9f3316c62b6fd616a64e96a447.

Then, the following functions were changed to deal with the low protein content. In the end this narrows down to changing the calculation of the UB of the protein_pool exchange reaction. These changes are expected to have limited effect on models where protein_pool is not limiting, but can be crucial to get low protein content models to function:

- `generate_protModels`:
  - Calculate sample-specific `f`-factor (see `measureAbundance` above), before filtering of the proteomics data. Even though individual proteins might have quite high variability, this does not affect the `f` calculation to much as it considers _all_ proteins, while this approach does ensure the highest coverage of protein, resulting in more representative `f`-factor. Commit: 66def23de4aa7145b90465dfbbeabea80e1422a6.
  - Before running `constrainEnzymes`, the individual measured protein abundances can be scaled to reach their minimum required usage. Overall, this increases the measured part of `Ptot`, thereby decreasing the unmeasured part of `Ptot`. Instead, `Ptot` should scale by the same ratio as the total measured part is scaled. To do this, the scaled abundances are compared to the abundances when there were first filtered. Commit: c70b766f54c8597fb88b0910de654270d48fe7bc.
- `constrainEnzymes`: simplify calculation of protein_pool exchange UB. First determine the expected total enzyme concentration, by multiplying the sample-specific `f` factor and scaled `Ptot` from `generate_protModels`. Subtract `Pmeasured` from this to yield the remaining that should then draw from the protein_pool. Commit 2c05284c82a29f17cce7b37e3e45d2c7a5934b5d.

These changes were essential to get low-protein models to work, but could also be beneficial in more "normal" conditions.

### CODE REVIEW

**Comment by edkerk** [comment]
One additional proposed change (commit cba9961):
- `generate_protModels` should always re-scale the biomass and calculate GAM if the protein content in the data is not the same as in the model. Previously this was only done if the difference was >5%, but this is a relatively arbitrary cut-off, particularly in a function that makes multiple models, where it could be that some models had a difference of 5.1% and other 4.9%. It doesn't take so much to scale biomass, so always do this if the protein content does not already match.

Otherwise, the proposed changes are complete and bugs fixed.

**Review by BenjaSanchez — APPROVED** [review]
Thanks for the changes! The refactoring did not change in any way `ecYeastGEM` and `ecYeastGEM_batch`, and as expected it changed `ecYeastGEM_prot` a little bit, which I included in a commit I just pushed ([`768c8ad`](https://github.com/SysBioChalmers/GECKO/pull/130/commits/768c8ad9ef8e4192e6e0f78d4ec1ff39f503b65e)). I only found one issue that I mention below, after addressing it this is ready to be merged!

**Review by IVANDOMENZAIN — COMMENTED** [review]
Wait before merge, there is a particular issue that I would like to address here, however I'm still on meetings and cannot write more right now, I'll comment on this PR later today.

File: geckomat/limit_proteins/constrainEnzymes.m [hunk 1]
```diff
@@ -83,12 +86,11 @@
...
-    [fn,~] = measureAbundance(model.enzymes(~measured));
-    fm     = Pmeasured/Ptot;
-    f      = fn/(1-fm);
-    %Discount measured mass from global constrain:
-    fs = (Ptot - Pmeasured)/Pbase*f*sigma;
+    %Expected total enzyme concentration
+    enzymeConc=Ptot*f;
+    %Non-measured part will be pooled
+    Ppool=enzymeConc-Pmeasured;
+    fs=Ppool/Pbase;
```
[the remark this answers is not in the document]
Comment (edkerk) [reply geckomat/limit_proteins/constrainEnzymes.m]: You're right, I submitted your suggestion.

**Comment by edkerk** [comment]
After @IVANDOMENZAIN fixed a mistake in #141, I now realize that it might not make sense to use condition-specific `f`-factor.

The `f`-factor is indicating how much of the total protein content is taken up by proteins that are included in `model.enzymes`. This can now be calculated from the condition-specific proteomics data. This `f`-factor is then used to define the UB of the `prot_pool_exchange` reaction:
- If total protein is 0.5 g/gDCW and typically 50% of all proteins are enzymes (i.e. included in `model.enzymes`, adding up to 0.25 g/gDCW), then in a proteomics dataset where the `model.enzymes` only added up to 0.2 g/gDCW, then the UB of the protein pool is 0.05 (`Ptot*f - Pmeasured` = 0.5*0.5 - 0.20).
- However, here I propose to calculate a condition-specific `f`-factor, which in this case would be 0.4 (of 0.5 g/gDCW total protein, 40% (=0.2 g/gDCW) is in `model.enzymes`), the calculation becomes `Ptot*f - Pmeasured` = 0.5*0.4 - 0.20 = 0. So only the measured proteins are allowed to be used, and no flux is allowed to go through the `prot_pool_exchange` reaction. The model fails, and `updateProtPool` then rather leaves the `prot_pool_exchange` reaction unconstrained.

However, this problem only arrises if `Ptot = sum(abundances)`: the sum of all measured protein abundances equals the total protein content. This assumes that all proteins are measured. Is that a fair assumption? If anything, **there** is probably where the problem lies in the models that I've been investigating. This problem therefore did not arise with the code that was used to test #141, where in the 'low' condition `Ptot` is 0.4228 while `sum(abundances)` is 0.381.

**Comment by mihai-sysbio** [comment]
>  This assumes that all proteins are measured. Is that a fair assumption?

This may be a stupid question: is it known how the ecModel is affected when a several `f`-factors are tried out? and to what extent the differences in the resulting models are proportional?

**Comment by edkerk** [comment]
> > This assumes that all proteins are measured. Is that a fair assumption?
> 
> This may be a stupid question: is it known how the ecModel is affected when a several `f`-factors are tried out? and to what extent the differences in the resulting models are proportional?

`f`-factor is just used to set the protein pool exchange upper bound, so varying its value is similar to simulating different levels of total protein. In _S. cerevisiae_ a lower `f`-factor would mean that Crabtree effect occurs at lower growth rates, in proportion to the reduced `f`. Alternatively, the sigma factor would be fitted higher to compensate for the lower `f`-factor.

I have been told, based on experience, that in previous _S. cerevisiae_ proteomics datasets about 80% of the total protein content was detected in the proteomics. This 80% coverage is now used in the non-_S. cerevisiae_ project as a rough ballpark figure, and solves the issue of too constraining protein pool exchange.

**Comment by BenjaSanchez** [comment]
@IVANDOMENZAIN after the changes in #141, is this PR ready to merge? All good on my side :)

**Comment by mihai-sysbio** [comment]
It would have been nice with a comment as to why the work in this PR has been discarded.

**Comment by edkerk** [comment]
I removed a few pointless branches on my PC (those that were linked to intermediate commits that have since been incorporated). It must have looked like this was the case for this branch, but I didn't notice that it actually had an PR standing open. Regardless, I do not intend to commit more time into this PR, but rather focus on the ongoing refactoring and perhaps consider the issues that were raised here.

**Pull request closed without merging**